targets/TARGET_STM/TARGET_STM32H7/i2c_device.c

```diff
@@ -0,0 +1,370 @@
+      tI2CCLK < (tLOW - tfilters) / 4 and tI2CCLK < tHIGH */
+      if ((tscl_l > I2C_Charac[I2C_speed].lscl_min) && (ti2cclk < ((tscl_l - tafdel_min - dnf_delay) / 4U)))
+      {
+        for (sclh = 0; sclh < I2C_SCLH_MAX; sclh++)
+        {
+          /* tHIGH(min) <= tAF(min) + tDNF + 2 x tI2CCLK + [(SCLH+1) x tPRESC] */
+          uint32_t tscl_h = tafdel_min + dnf_delay + (2U * ti2cclk) + ((sclh + 1U) * tpresc);
+
+          /* tSCL = tf + tLOW + tr + tHIGH */
+          uint32_t tscl = tscl_l + tscl_h + I2C_Charac[I2C_speed].trise + I2C_Charac[I2C_speed].tfall;
+
+          if ((tscl >= clk_min) && (tscl <= clk_max) && (tscl_h >= I2C_Charac[I2C_speed].hscl_min) && (ti2cclk < tscl_h))
+          {
+            int32_t error = (int32_t)tscl - (int32_t)ti2cspeed;
+
+            if (error < 0)
+            {
+              error = -error;
+            }
+
+            /* look for the timings with the lowest clock error */
+            if ((uint32_t)error < prev_error)
+            {
+              prev_error = (uint32_t)error;
+              I2c_valid_timing[count].scll = scll;
+              I2c_valid_timing[count].sclh = sclh;
+              ret = count;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  return ret;
+}
+
+/**
+  * @brief  Compute I2C timing according current I2C clock source and required I2C clock.
+  * @param  clock_src_freq I2C clock source in Hz.
+  * @param  i2c_freq Required I2C clock in Hz.
+  * @retval I2C timing or 0 in case of error.
+  */
+static uint32_t I2C_ComputeTiming(uint32_t clock_src_freq, uint32_t i2c_freq)
+{
+  uint32_t ret = 0;
+  uint32_t speed;
+  uint32_t idx;
+
+
+  if((clock_src_freq != 0U) && (i2c_freq != 0U))
+  {
+    for ( speed = 0 ; speed <=  (uint32_t)I2C_SPEED_FREQ_FAST_PLUS ; speed++)
+    {
+      if ((i2c_freq >= I2C_Charac[speed].freq_min) &&
+          (i2c_freq <= I2C_Charac[speed].freq_max))
+      {
+        I2C_Compute_PRESC_SCLDEL_SDADEL(clock_src_freq, speed);
+        idx = I2C_Compute_SCLL_SCLH(clock_src_freq, speed);
+
+        if (idx < I2C_VALID_TIMING_NBR)
+        {
+          ret = ((I2c_valid_timing[idx].presc  & 0x0FU) << 28) |\
+                ((I2c_valid_timing[idx].tscldel & 0x0FU) << 20) |\
+                ((I2c_valid_timing[idx].tsdadel & 0x0FU) << 16) |\
+                ((I2c_valid_timing[idx].sclh & 0xFFU) << 8) |\
+                ((I2c_valid_timing[idx].scll & 0xFFU) << 0);
+        }
+        break;
+      }
+    }
+  }
+
+  return ret;
+}
+/**
+  * @}
+  */
+
+/** @defgroup I2C_DEVICE_Exported_Functions I2C_DEVICE Exported Functions
+  * @{
+  */
+/**
+ * @brief  Provide the suitable timing depending on requested frequency
+ * @param  hz Required I2C clock in Hz.
+ * @retval I2C timing or 0 in case of error.
+ */
+uint32_t get_i2c_timing(int hz)
+{
+  uint32_t tim;
+
+  I2c_valid_timing_nbr = 0;
+
+  /* we will use D2PCLK1 to calculate I2C timings */
+  MBED_ASSERT(RCC_I2C1CLKSOURCE_D2PCLK1 ==__HAL_RCC_GET_I2C1_SOURCE());
```

Comment:
This is strange as H747-DISCO is suing I2C4 for D14/D15 pins.

NUCLEO H743 is using I2C1...

Reply:
When i wrote this line i was making wrong assumptions and it is completely useless. On DISCO I2C1_SOURCE is actually D2PCLK1 so we never hit the assert, moreover as you wrote I2C4 is used for the tests... On DISCO tests are ok because PCLK1_Frequency, PCLK2_Frequency, PCLK3_Frequency and PCLK4_Frequency have all the same value.

On the other side I2C1 instance is used for running tests on the NUCLEO and we doesn't hit the assert too, this means we are using the correct clock source to compute timings... so i thinks there is something else i'm missing to make things work on the NUCLEO.